### PR TITLE
session fixes

### PR DESCRIPTION
**Pull request by tannernelson**
This makes some updates to Session so that we don't create them unnecessarily.

If you want to start a new session, you do so explicitly like 

```swift
if request.session == nil {
    //create a new session
    let session = Session(driver: app.session)
    session["name"] = "Vapor"

    request.session = session
}
```

We could create some sort of helper method on `app` to create sessions so that you don't have to remember to pass the `app.session` driver.

```swift
app.makeSession()
```

It also fixes a bug where header keys from the new Socket class weren't getting converted to lowercase so `headers["cookie"]` was not finding the header `"Cookie"`. We should write a test to make sure this doesn't break again.

@qutheory/vapor Let me know what you think.

### CODE REVIEW

**Comment by ketzusaka** [comment]
I don't think there's any issue creating a session for each request. It doesn't ask the driver to do any storage operations until you attempt to use it.

I guess my biggest peeve is now each request that may want to store a session has way more responsibility. Maybe a static var on SessionMiddleware to auto create them, if this is something you really want?

**Comment by tannernelson** [comment]
I think how people are going to use it is have some `AuthMiddleware` or something that will check for a session to make sure the user is logged in. Something like this:

```swift
guard let session = request.session else {
  throw Error.NotLoggedIn
}
```

And in the login route something like this 

```swift
guard usernameAndPasswordAreCorrect else {
  throw Error.InvalidCredentials
}
request.session = Session()
```

This way you're only having Vapor send a `vapor-session` cookie if you actually intend to use it.

File: Sources/Vapor/Request/Request.swift [hunk 1]
```diff
@@ -42,7 +42,7 @@ public class Request {
     public var parameters: [String: String] = [:]
     
     ///Server stored information related from session cookie.
-    public internal(set) var session: Session?
+    public var session: Session?
```
Comment (ketzusaka) [inline Sources/Vapor/Request/Request.swift]: Could this go back to internally settable?

File: Sources/Vapor/Session/Session.swift [hunk 2]
```diff
@@ -3,13 +3,22 @@ public class Session {
 
     public let identifier: String
     var driver: SessionDriver
+    public var enabled: Bool
+
+    public init(driver: SessionDriver) {
```
Comment (ketzusaka) [inline Sources/Vapor/Session/Session.swift]: And this internal?

Comment (tannernelson) [inline Sources/Vapor/Session/Session.swift]: The user might want to override `SessionMiddleware` with their own solution.

Comment (ketzusaka) [inline Sources/Vapor/Session/Session.swift]: fair enough

**Comment by ketzusaka** [comment]
:+1:

**Pull request merged**